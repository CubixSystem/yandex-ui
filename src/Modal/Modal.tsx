import React, { FC, MouseEventHandler, Ref, RefObject, useRef } from 'react';
import { cn } from '@bem-react/classname';

import { usePreventScroll } from '../usePreventScroll';
import { useFocusTrap } from '../useFocusTrap';
import { useOverlay, OnClose } from '../useOverlay';
import { PortalExtendableProps, Portal } from '../Portal';
import './Modal.css';

interface FocusTrapProps {
    /**
     * Включить ловушку фокуса
     *
     * @default true
     */
    trapFocus?: boolean;

    /**
     * Выставить фокус на первый интерактивный элемент
     * в пределах модального окна
     *
     * @default true
     */
    autoFocus?: boolean;

    /**
     * Вернуть фокус на элемент после закрытия,
     * который вызвал модальное окно
     *
     * @default true
     */
    restoreFocus?: boolean;

    /**
     * Ссылка на элемент, на который будет возвращен
     * фокус после закрытия
     */
    restoreFocusRef?: RefObject<HTMLElement>;
}

export interface IModalProps extends PortalExtendableProps, FocusTrapProps {
    /**
     * id элемента, который описывает это модальное окно. NVDA проговорит этот элемент, когда окно откроется.
     * Несколько id можно указать через пробел.
     */
    ariaLabelledBy?: string;

    /**
     * Выравнивание контента по вертикали
     *
     * @default 'middle'
     */
    contentVerticalAlign?: 'top' | 'middle' | 'bottom';

    /**
     * Добавляет анимацию при открытии модального окна.
     *
     * @default true
     */
    hasAnimation?: boolean;

    /**
     * Блокирует прокрутку на теле документа, при открытом модальном окне.
     *
     * @default true
     */
    preventBodyScroll?: boolean;

    /**
     * Дополнительный класс у корневого DOM-элемента
     */
    className?: string;

    /**
     * Обработчик, вызываемый при срабатывании события click
     */
    onClick?: MouseEventHandler<HTMLDivElement>;

    /**
     * Обработчик, вызывающийся после нажатия на клавишу esc,
     * либо мышкой на область вне контейнера
     */
    onClose?: OnClose;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: Ref<HTMLDivElement>;

    /**
     * Задает слой `z-index`
     */
    zIndex?: number;
}

export const cnModal = cn('Modal');

/**
 * Используется для создания всплывающих модальных окон.
 * @param {IModalProps} props
 */
export const Modal: FC<IModalProps> = ({
    children,
    className,
    contentVerticalAlign: align = 'middle',
    hasAnimation = true,
    innerRef,
    keepMounted,
    onClose,
    preventBodyScroll = true,
    scope,
    visible,
    zIndex,
    trapFocus = true,
    autoFocus = true,
    restoreFocus = true,
    restoreFocusRef,
    ariaLabelledBy,
    ...props
}) => {
    const contentRef = useRef<HTMLDivElement>(null);

    usePreventScroll({ enabled: preventBodyScroll && visible });
    useFocusTrap({
        enabled: trapFocus && visible,
        scopeRef: contentRef,
        autoFocus,
        restoreFocus,
        restoreFocusRef,
    });
    useOverlay({
        visible,
        onClose,
        essentialRefs: [contentRef],
        unsafe_strategy: 'pressup',
    });

    return (
        <Portal scope={scope} visible={visible} keepMounted={keepMounted}>
            <div
                {...props}
                ref={innerRef}
                className={cnModal({ visible, hasAnimation }, [className])}
                style={{ zIndex }}
            >
                <div className={cnModal('Overlay')} />
                <div className={cnModal('Wrapper')}>
                    <div className={cnModal('Table')}>
                        <div className={cnModal('Cell', { align })}>
                            <div ref={contentRef} className={cnModal('Content')} tabIndex={-1} role="dialog" aria-labelledby={ariaLabelledBy} aria-modal>
                                {children}
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </Portal>
    );
};

Modal.displayName = cnModal();
