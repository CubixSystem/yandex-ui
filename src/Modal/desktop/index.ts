export * from '../Modal@desktop';
export * from '../_theme/Modal_theme_normal@desktop';
