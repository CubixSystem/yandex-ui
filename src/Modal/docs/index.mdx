---
id: Modal
path: /components/modal

title: Modal
description: Компонент для создания меню.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Modal
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Modal.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fsurface-modal-desktop--playground

tabs:
 - Examples
 - Code
 - Notes
---
