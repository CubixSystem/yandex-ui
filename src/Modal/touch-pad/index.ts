export * from '../Modal@touch-pad';
export * from '../_theme/Modal_theme_normal';
