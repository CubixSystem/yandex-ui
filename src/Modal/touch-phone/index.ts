export * from '../Modal@touch-phone';
export * from '../_theme/Modal_theme_normal';
