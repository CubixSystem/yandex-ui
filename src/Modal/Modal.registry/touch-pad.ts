import { Registry } from '@bem-react/di';

import { cnModal } from '../Modal';
import { Popup } from '../../Popup/Popup@touch-pad';

export const modalRegistry = new Registry({ id: cnModal() });

modalRegistry.set('Popup', Popup);
