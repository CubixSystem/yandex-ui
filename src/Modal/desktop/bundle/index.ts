import { compose } from '@bem-react/core';

// base
import { Modal as ModalDesktop } from '../../Modal';
// _theme
import { withThemeNormal } from '../../_theme/Modal_theme_normal@desktop';

export * from '../../Modal';

export const Modal = compose(withThemeNormal)(ModalDesktop);
