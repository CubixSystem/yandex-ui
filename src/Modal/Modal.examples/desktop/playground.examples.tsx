import React, { createRef, useState } from 'react';
import { withKnobs } from '@storybook/addon-knobs';

import { Button } from '../../../Button/Button.bundle/desktop';
import { Modal } from '../../Modal.bundle/desktop';
import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: [withKnobs, ...createDecorators({ scope: 'desktop' })],
    parameters,
};

const scopeRef = createRef<HTMLDivElement>();

export const Playground = () => {
    const [visible, setVisible] = useState(false);
    return (
        <div ref={scopeRef} style={{ height: 72 }}>
            <Button view="default" size="m" onClick={() => setVisible(true)}>
                Открыть
            </Button>
            <Modal theme="normal" scope={scopeRef} visible={visible} onClose={() => setVisible(false)}>
                <div style={{ padding: 16, fontFamily: 'var(--control-font-family)', width: 400 }}>
                    <div style={{ marginBottom: 16 }}>
                        Общедоступная многоязычная универсальная интернет-энциклопедия со свободным контентом.
                    </div>
                    <div style={{ display: 'flex', justifyContent: 'flex-end' }}>
                        <Button view="clear" size="m" onClick={() => setVisible(false)}>
                            Отменить
                        </Button>
                        <Button view="default" size="m" onClick={() => setVisible(false)}>
                            Хорошо
                        </Button>
                    </div>
                </div>
            </Modal>
        </div>
    );
};

Playground.story = {
    name: 'playground',
};
