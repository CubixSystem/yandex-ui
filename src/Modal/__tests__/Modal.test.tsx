import React, { createRef } from 'react';

import { createClientRender } from '../../internal/testing';
import { Modal as ModalCommon } from '../Modal';
import { Modal as ModalDesktop } from '../desktop';
import { Modal as ModalTouchPad } from '../touch-phone';
import { Modal as ModalTouchPhone } from '../touch-pad';

const platforms = [
    ['desktop', ModalDesktop],
    ['touch-pad', ModalTouchPad],
    ['touch-phone', ModalTouchPhone],
] as const;

describe.each(platforms)('Modal@%s', (_platform, Modal) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(
            <Modal visible className="mix" scope="inplace">
                content
            </Modal>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен устанавливать ссылку на корневой DOM элемент', () => {
        const innerRef = createRef<HTMLDivElement>();
        render(<Modal visible innerRef={innerRef} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен быть установлен корректный displayName', () => {
        expect(ModalCommon.displayName).toBe('Modal');
    });
});
