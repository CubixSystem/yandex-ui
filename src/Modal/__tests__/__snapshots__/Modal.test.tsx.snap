// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Modal Modal_visible Modal_hasAnimation mix"
>
  <div
    class="Modal-Overlay"
  />
  <div
    class="Modal-Wrapper"
  >
    <div
      class="Modal-Table"
    >
      <div
        class="Modal-Cell Modal-Cell_align_middle"
      >
        <span
          aria-hidden="true"
          data-focus-trap-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="1"
        />
        <div
          aria-modal="true"
          class="Modal-Content"
          role="dialog"
          tabindex="-1"
        >
          content
        </div>
        <span
          aria-hidden="true"
          data-focus-trap-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Modal Modal_visible Modal_hasAnimation mix"
>
  <div
    class="Modal-Overlay"
  />
  <div
    class="Modal-Wrapper"
  >
    <div
      class="Modal-Table"
    >
      <div
        class="Modal-Cell Modal-Cell_align_middle"
      >
        <span
          aria-hidden="true"
          data-focus-trap-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="1"
        />
        <div
          aria-modal="true"
          class="Modal-Content"
          role="dialog"
          tabindex="-1"
        >
          content
        </div>
        <span
          aria-hidden="true"
          data-focus-trap-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Modal Modal_visible Modal_hasAnimation mix"
>
  <div
    class="Modal-Overlay"
  />
  <div
    class="Modal-Wrapper"
  >
    <div
      class="Modal-Table"
    >
      <div
        class="Modal-Cell Modal-Cell_align_middle"
      >
        <span
          aria-hidden="true"
          data-focus-trap-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="1"
        />
        <div
          aria-modal="true"
          class="Modal-Content"
          role="dialog"
          tabindex="-1"
        >
          content
        </div>
        <span
          aria-hidden="true"
          data-focus-trap-guard="true"
          style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</div>
`;
